File: Algo/LeadToOffer/68_NumToStringCases.cpp
//recursive
class Solution {
public:
    int translateNum(int num) {
        if(num < 10)
            return 1;
        if(num < 26)
            return 2;
        if(num < 100)
            return 1;

        return translateNum(num / 10) + translateNum(num / 100) * (translateNum(num % 100) - 1);
    }
};

// dp
class Solution {
public:
    int translateNum(int num) {
        if(num < 10)
            return 1;
        if(num < 26)
            return 2;
        if(num < 100)
            return 1;

        int f1 = 1; int f2 = 1; int f;
        string v = to_string(num);
        int digit = v.length();
        for(int i = 2; i <= digit; i++){
            int k = stoi(v.substr(i-2, 2));
            f = f1 + f2 * (k >= 10 && k <= 25);

            f2 = f1;
            f1 = f;
        }
        return f;
    }
};

File: Algo/LeadToOffer/68_NumToStringCases.py
class Solution(object):
    def translateNum(self, num):
        """
        :type num: int
        :rtype: int
        """
        s = str(num)
        n = len(s)
        dp = [1] * (n+1)
        dp[1] = 1
        if n == 1:
            return 1

        dp[2] = 1 if int(s[:2]) > 25 else 2
        for i in range(3, n+1):     # must consider 02 case
            dp[i] = dp[i-1] + dp[i-2] if 10 <= int(s[i-2:i]) <= 25 else dp[i-1]

        return dp[-1]